### DISCUSSION TITLE
Extract schemas for PDFs?

### DISCUSSION THREAD

**Discussion by mdfederici** [firecrawl/firecrawl#806] [Q&A]
Do extracts not work with PDFs? I can use scrape_url to scrape a webpage and include an 'extract' schema for the LLM to populate. I can use scrape_url to scrape a pdf without specifying an extract. But if i include an extract with a pdf I seem to get a bad request response. I would have thought that once the pdf is converted to markdown it would go through the same process/pipes for extracting a custom schema. Thanks in advance for any tips.

**Comment by Adit-Jain-srm** [comment]
Extract with PDFs does work, but you need to enable PDF parsing explicitly. The key is to include `parsePDF: true` (which is the default) AND use the `extract` format:

```python
from firecrawl import FirecrawlApp

app = FirecrawlApp(api_key='your-key')

result = app.scrape_url(
    'https://example.com/document.pdf',
    params={
        'formats': ['extract'],
        'extract': {
            'schema': {
                'type': 'object',
                'properties': {
                    'title': {'type': 'string'},
                    'author': {'type': 'string'},
                    'key_findings': {
                        'type': 'array',
                        'items': {'type': 'string'}
                    }
                }
            }
        }
    }
)

print(result.get('extract'))
```

If you're getting a bad request, check:
1. Make sure `formats` includes `'extract'` (not just `'markdown'`)
2. Your `extract.schema` must be valid JSON Schema
3. The PDF must be accessible (not behind auth that Firecrawl can't handle)

The pipeline is: PDF → text extraction → markdown conversion → LLM extraction with your schema. It works the same as HTML pages once the PDF is parsed.
